File: slurp/slurp.py
#!/usr/bin/env python
#Copyright (c) 2013, Charles Duyk
#All rights reserved.
#
#Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
#
#Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
#Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
#THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
import argparse
import os
import sys
import requests
import time

DEFAULT_MESSAGE_COUNT = 100
DEFAULT_TOKEN_FILE_NAME = "token"
DEFAULT_THREAD_ID_FILE_NAME = "threadID"

tokenHelpText  = """A Facebook access token or a path to a file containing the access token on the first line. The token must have the 'mailbox_read' permission. If absent, the script searches the current directory for a file named '%s'""" % (DEFAULT_TOKEN_FILE_NAME)
threadHelpText = """ThreadID of the thread or path to a file the thread ID on the first line. If absent, the script searches the current directory for a file named '%s'""" % (DEFAULT_THREAD_ID_FILE_NAME)
numMessageHelpText = """Number of messages to retrieve. Default: %d""" % (DEFAULT_MESSAGE_COUNT)
outputHelpText="""File to write the messages to. If absent, writes to stdout"""

progress = False

class Message(object):
	def __init__(self, time, sender, text):
		super(Message, self).__init__()
		self._time = time
		self._sender = sender
		self._text = text
	
	def __str__(self):
		timeString = time.ctime(self._time)
		return u"(%s) %s: %s" % (timeString, self._sender, self._text)
	
def parseArgs():
	parser = argparse.ArgumentParser()
	parser.add_argument("-t", "--token", dest="tokenOrPath", metavar="TOKEN", default=DEFAULT_TOKEN_FILE_NAME, help=tokenHelpText);
	parser.add_argument("-i", "--threadID", dest="idOrPath", metavar="ID", default=DEFAULT_THREAD_ID_FILE_NAME, help=threadHelpText)
	parser.add_argument("-n", "--num-messages", required=False, type=int, default=DEFAULT_MESSAGE_COUNT, help=numMessageHelpText)
	parser.add_argument("-o", "--output", required=False, metavar="PATH", help=outputHelpText)
	parser.add_argument("-p", "--progress", action="store_true", help="Print progress updates to stderr")
	return parser.parse_args()

def getKey(keyOrPath):
	key = ""
	if os.path.exists(keyOrPath):
		with open(keyOrPath) as keyFile:
			key = keyFile.readline().strip()
	else:
		key = keyOrPath
	return key

def slurpMessages(threadID, token, numMessages):
	url = "https://graph.facebook.com/%s/comments" % (threadID)
	messages = []
	params = {}
	params["access_token"] = token
	params["date_format"] = 'U'
	i = 0
	while i < numMessages:
		if progress:
			sys.stderr.write("Slurping message %d\n" % i)
		response = requests.get(url, params=params)
		json = response.json()
		try:
			recieved_messages_json = json["data"]
		except KeyError as e:
			sys.stderr.write("Couldn't parse JSON response (%s) %s\n" % (e.message, json))
			break
		for message_json in recieved_messages_json:
			try:
				time = int(message_json["created_time"])
				sender = message_json["from"]["name"]
				text = message_json["message"]
				message = Message(time, sender, text)
				messages.append(message)
			except KeyError as e:
				sys.stderr.write("Coudln't parse message %d (%s) %s\n" % (i, e, message_json))
			i += 1
			if i >= numMessages:
				break
		url = json["paging"]["next"]
	return messages

def main():
	global progress
	args = parseArgs()
	progress = args.progress
	numMessages = args.num_messages
	token = getKey(args.tokenOrPath)
	threadID = getKey(args.idOrPath)
	outputFile = sys.stdout
	if args.output is not None:
		outputPath = args.output
		outputFile = open(outputPath, mode='w')
	messages = slurpMessages(threadID, token, numMessages)
	messages.sort(cmp=lambda x, y: cmp(x._time, y._time))
	output = u"\n".join(map(unicode, messages)).encode('utf-8')
	outputFile.write(output)
	outputFile.write("\n")
	outputFile.close()

if __name__ == "__main__":
	main()

File: goSlurp/slurp.go
package main

import "fmt"
import "log"
import "flag"
import "os"
import "io"
import "io/ioutil"
import "bufio"
import "bytes"
import "net/url"
import "net/http"
import "encoding/json"

type User struct {
	Name string
	Id int
}


type Message struct {
	Id string
	Timestamp int 
	User User
	Text string
}

const DEFAULT_TOKEN_PATH = "token"
const DEFAULT_THREADID_PATH = "id"

func init() {
	log.SetFlags(0)
}

func readFirstLine(path string) (string, error) {
	var line []byte
	var isPrefix bool
	var err error
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()
	reader := bufio.NewReader(file)
	slices := make([][]byte, 0)
	isPrefix = true
	for isPrefix {
		line, isPrefix, err = reader.ReadLine()
		if err != nil {
			break;
		}
		slices = append(slices, line)
	}
	lineBytes := bytes.Join(slices, nil)
	return string(lineBytes), err
}

func makeURL(token, id string) string {
	params := url.Values{}
	params.Add("access_token", token)
	params.Add("date_format", "U")
	paramString := params.Encode()
	url := fmt.Sprintf("https://graph.facebook.com/%s/comments?%s", id, paramString)
	return url
}

func ParseUser(json map[string]interface{}) User {
	u := User{}
	u.Name = json["name"].(string)
	u.Id = int(json["id"].(float64))
	return u
}

func ParseMessage(json map[string]interface{}) Message {
	m := Message{}
	m.Id = json["id"].(string)
	m.Timestamp = int(json["created_time"].(float64))
	m.User = ParseUser(json["from"].(map[string]interface{}))
	m.Text = json["message"].(string)
	return m
}

func parseMessages(body io.ReadCloser) ([]Message, error) {
	defer body.Close()
	var dest map[string]interface{}
	messages := make([]Message, 0, 25)
	bytes, _ := ioutil.ReadAll(body)
	json.Unmarshal(bytes, &dest)
	messageMaps := dest["data"].([]interface{})
	for _, v := range messageMaps {
		messages = append(messages, ParseMessage(v.(map[string]interface{})))
	}
	return nil, nil
}

func fetchMessages(url string, numMessages uint) ([]Message, error) {
	messages := make([]Message, 0, numMessages)
	response, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	parsed, err := parseMessages(response.Body)
	messages = append(messages, parsed...)
	return messages, err
}

func main() {
	var tokenPath string
	var idPath string
	flag.StringVar(&tokenPath, "token", DEFAULT_TOKEN_PATH, "Path to a file containing the Facebook access token on the first line.")
	flag.StringVar(&idPath, "id", DEFAULT_THREADID_PATH, "Path to a file containing the thread ID on the first line.")
	flag.Parse()
	token, err := readFirstLine(tokenPath)
	if err != nil {
		log.Fatal("Could not read token:", err)
	}
	id, err := readFirstLine(idPath)
	if err != nil {
		log.Fatalln("Could not read id:", err)
	}
	url := makeURL(token, id)
	messages, err := fetchMessages(url, 5)
	fmt.Println(messages)
}
